### DISCUSSION TITLE
os error 32?

### DISCUSSION THREAD

**Discussion by daranzolin** [quarto-dev/quarto-cli#10775] [Q&A]
### Description

I have been running the same Quarto reports for over a year, and today I am suddenly getting `Error: The process cannot access the file because it is being used by another process. (os error 32)`. I haven't changed anything. It's a fairly complex project within a targets pipeline, so before I try and create a reprex, can someone help me brainstorm possible reasons? I haven't updated Quarto, changed the code, or any file paths, etc.

**Comment by cscheid** [comment]
Some other process is attempting to open the file at the same time, and Windows doesn't like that. We haven't changed anything about that on our side either - maybe your OS upgrade caused something like anti-virus scanners to be more active?

**Reply by daranzolin** [reply]
Interesting, other Quarto reports are rendering, copying, and moving fine to the same locations, its just this specific batch...must be something unique about them, I'll have to dig deeper, thanks!

**Reply by mcanouil** [reply]
I am guessing your are using a cloud drive here.
This is a known issue reported many times already.

**Reply by daranzolin** [reply]
@mcanouil I saw those reports, and wondered if a stakeholder was trying to sync this shared report folder somehow, but I get the same errors when rendering directly to my local computer, and I do not have DropBox or OneDrive or anything like that.

**Reply by mcanouil** [reply]
Thanks for ruling this out as it was the most common case (also with os error 1920).

- https://github.com/quarto-dev/quarto-cli/issues/4670

**Reply by mcanouil** [reply]
Since you stated the issue happen on your local computer hard drive, would you mind try with a simple document to reproduce? In case it's something from your document.


---

<table>
<tr><th>R</th><th>Python</th></tr>
<tr><td>

`````md
````qmd
---
title: "Reproducible Quarto Document"
format: html
engine: knitr
---

This is a reproducible Quarto document.

{{< lipsum 1 >}}

```{r}
x <- c(1, 2, 3, 4, 5)
y <- c(1, 4, 9, 16, 25)

plot(x, y)
```

![An image]({{< placeholder 600 400 >}}){#fig-placeholder}

{{< lipsum 1 >}}

The end after @fig-placeholder.
````
`````

</td><td>

`````md
````qmd
---
title: "Reproducible Quarto Document"
format: html
engine: jupyter
---

This is a reproducible Quarto document.

{{< lipsum 1 >}}

```{python}
import matplotlib.pyplot as plt

x = [1, 2, 3, 4, 5]
y = [1, 4, 9, 16, 25]

plt.plot(x, y)
plt.show()
```

![An image]({{< placeholder 600 400 >}}){#fig-placeholder}

{{< lipsum 1 >}}

The end after @fig-placeholder.
````
`````

</td></tr>
</table>

**Reply by daranzolin** [reply]
This renders fine. If I had to guess, it's related to the presence of nested content via `knitr::knit_expand()` and `knitr::knit_child()`, since the only reports that get this error have *lots* of child documents.

**Reply by mcanouil** [reply]
Unfortunately, without the means to reproduce on our side with a small and simple example, it's going to be very hard to help.

**Reply by daranzolin** [reply]
When I remove the chunks (or just `eval=FALSE`) with `knitr::expand()` and `knitr::knit_child()` the reports render.

**Comment by daranzolin** [comment]
Moving this outside the replies. When I try and execute a qmd with `knitr::knir_expand()` and `knitr::knit_child()` I get the following error:

```
Error: The process cannot access the file because it is being used by another process. (os error 32), remove 'C:\<redacted>\Documents\Projects\<redacted>_files\execute-results'
    at Object.removeSync (deno:runtime/js/30_fs.js:157:9)
    at removeIfExists (file:///C:/<redacted>~1/AppData/Local/Programs/Quarto/bin/quarto.js:5681:14)
    at removeFreezeResults (file:///C:/<redacted>~1/AppData/Local/Programs/Quarto/bin/quarto.js:86413:5)
    at renderExecute (file:///C:/<redacted>~1/AppData/Local/Programs/Quarto/bin/quarto.js:86501:9)
    at async renderFiles (file:///<redacted>~1/AppData/Local/Programs/Quarto/bin/quarto.js:86593:51)
    at async renderProject (file:///<redacted>~1/AppData/Local/Programs/Quarto/bin/quarto.js:86800:25)
    at async Command.fn (file:///C:/<redacted>~1/AppData/Local/Programs/Quarto/bin/quarto.js:90856:32)
    at async Command.execute (file:///C:/<redacted>~1/AppData/Local/Programs/Quarto/bin/quarto.js:8437:13)
    at async quarto (file:///C:/<redacted>~1/AppData/Local/Programs/Quarto/bin/quarto.js:127545:5)
    at async file:///C:/<redacted>~1/AppData/Local/Programs/Quarto/bin/quarto.js:127563:9
```

**Reply by cscheid** [reply]
@daranzolin Can you give us a small .qmd file (as small as you can make it) with kntr_expand  and knit_child that consistently triggers this issue for you?

**Reply by daranzolin** [reply]
That's a tall order--the child documents themselves have child documents, and those child documents have child documents, three-to-four levels deep. At each level a different [{reactable} table](https://glin.github.io/reactable/) is generated, along with [{downloadthis} buttons](https://github.com/shafayetShafee/downloadthis) linking to xlsx files created beforehand. It's a monstrous project that probably needs to be simplified, but had worked well up until now. I'll see what I can do later this week.

**Reply by cscheid** [reply]
Sounds good, and thanks. Unfortunately, there isn't much we can do until we can try it on our own side.

**Reply by cderv** [reply]
Thanks for sharing the stack trace. It seems related to freezing process. Somehow, the `execute-results` can't be removed. 

@cscheid it makes me thing about the problem I am trying to resolve in this PR:
- https://github.com/quarto-dev/quarto-cli/pull/10629

OS locking conflict while Deno is trying to remove a directory. It happens for me sometimes with `mediabag` in VSCODE context. 

So I am wondering if this is somewhat the same type of issue where OS is for whatever reason still on lock with the file while Deno is trying to remove it and can't.

**Reply by daranzolin** [reply]
I did some more testing, and I'm now pretty convinced it's related to the *quantity* of child content spawned at the deepest level in the parent-child hierarchy. When I reduce it to five groups, it renders every time. It fails at six groups every time. 

Edit: more likely it's related to either (1) removing the reactable tables from the child documents; or (2) the downloadthis extension is looking for files that perhaps don't exist.

**Reply by mcanouil** [reply]
> downloadthis extension

What do you mean here? The R package, the Quarto extension? Something else?

At some point, you'll have to provide concrete example of your setup.
I suggest you remove pieces one at a time to end up with something small that could be share.
We can keep guessing but it won't get us close to an actual solution (except if we are lucky).

I suggest you update this thread when you have something we can work on.

**Reply by daranzolin** [reply]
@mcanouil Understood. See above.
> At each level a different [{reactable} table](https://glin.github.io/reactable/) is generated, along with [{downloadthis} buttons](https://github.com/shafayetShafee/downloadthis)

**Comment by xx02al** [comment]
Hi
I am aware that pre-release builds are intended for testing purposes, and are not recommended for general use. Nevertheless, I thought it may help to make you aware of the following. With quarto v1.7.16 `quarto render` in the terminal errors as illustrated next (the path is shortened). I observed it on a windows machine.

```
Error occured during cleanup: Error: The process cannot access the file because it is being used by another process. (os error 32), remove 'C:\...\Temp\quarto-sessionid'
```

This occurs with Positron as well as on RStudio. I tried it an can confirm downgrading quarto to 1.6.42 helps have things working without any error - using version v1.7.16 has this issue.

Below are some reprex files with which I can observe it. Can you reproduce the issue, do you know a reason and how to fix it!?

_quarto.yml

```
project:
  type: book

book:
  title: "another-book"
  author: "Norah Jones"
  date: "2025-03-14"
  chapters:
    - index.qmd
    - intro.qmd
    - summary.qmd
    - references.qmd

bibliography: references.bib

format:
  html:
    theme:
      - cosmo
      - brand
  pdf:
    documentclass: scrreprt
```

index.qmd

``` qmd
# Preface {.unnumbered}

This is a Quarto book.

To learn more about Quarto books visit <https://quarto.org/docs/books>.
```

intro.qmd

``` qmd
# Introduction

This is a book created from markdown and executable code.

See @knuth84 for additional discussion of literate programming.
```

references.bib

```
@article{knuth84,
  author = {Knuth, Donald E.},
  title = {Literate Programming},
  year = {1984},
  issue_date = {May 1984},
  publisher = {Oxford University Press, Inc.},
  address = {USA},
  volume = {27},
  number = {2},
  issn = {0010-4620},
  url = {https://doi.org/10.1093/comjnl/27.2.97},
  doi = {10.1093/comjnl/27.2.97},
  journal = {Comput. J.},
  month = may,
  pages = {97–111},
  numpages = {15}
}
```

references.qmd

``` qmd
# References {.unnumbered}

::: {#refs}
:::
```

summary.qmd

``` qmd
# Summary

In summary, this book has no content whatsoever.
```

**Reply by mcanouil** [reply]
@xx02al Is you example `quarto create project book`? If not, please share as a Git repository (Gists are Git repositories if you don't want to create a "real" repository).

**Reply by mcanouil** [reply]
Linking again for reference:
- https://github.com/quarto-dev/quarto-cli/issues/4670
- https://github.com/quarto-dev/quarto-cli/discussions/10775#discussioncomment-10626369

**Reply by xx02al** [reply]
@mcanouil, yes, to have a reprex, I simply went into the command palette on positron Quarto: Create Project selected Book Project. I deleted the .png image, updated the book title, and corrected the date to a valid format.

**Reply by mcanouil** [reply]
Ok, thanks, so maybe you can update your example after double checking that the error still appears from `quarto create project book`?

Note that when sharing multiple files, asking for other to create the structure, create the files and finally copy/paste the content is quite tedious and time consuming. In this case, you also had to open each file to copy/paste content.
Since you are using Positron, it's quite easy to create a Git repository with all the files at the right place with the right content. It saves time for everyone.

This being said:
- Do you confirm that you are not working on a network/cloud drive with active syncing?

CC. @cderv the Windows expert (or the only Quarto team member using Windows^^)

**Reply by xx02al** [reply]
Totally agree on the copy/paste aspects, @mcanouil, I thought you were so familiar with it that you would immediately notice its basically `quarto create project book`... but I should have mentioned that and the small corrections I made before I showed the files (if showing them as I did).

I hereby confirm that the error also occurs with `quarto create project book` in the terminal. And also I am not working on a network/cloud drive with active syncing.

**Reply by cderv** [reply]
I believe this is directly related to a behavior I saw with some projects and with the dev version of Quattro due to the locking mechanism on Windows. 

I have tried to do some mitigation at
- https://github.com/quarto-dev/quarto-cli/pull/10629

Ultimately, this is a problem with Deno v1 and file locking on Windows while trying to remove it. When working in IDE, I believe those IDES are scanning or doing something else with the files while Quarto is rendering, and so when Quarto tries to clean, the file is locked. 

I'll try to check how 1.7.16 behaves on my side. 

@cscheid, if confirmed, we may need to reconsider #10629's idea to try to help with this error on Windows.

**Reply by cderv** [reply]
So I can reproduce with latest 1.7 version. 

It happens in 1.7.14 but not in 1.7.13. I'll check if we did something incorrectly, or just added a cleaning that triggers behavior seen in #10629

**Reply by cderv** [reply]
Found why. This is in fact a deep silent issue. Thanks a lot @xx02al !!

**Reply by cderv** [reply]
@xx02al this error will be fixed in next pre-release 1.7.18

**Reply by xx02al** [reply]
great! thanks @cderv for your efforts and also this info, I was just looking again 2, 3 hours ago to see what the situation might be

**Reply by andreashandel** [reply]
I seem to have the same problem with Quarto (1.7.32) through Positron. If I open a project/repo and do a 'quarto render' in the terminal, it works. If I then I try to run the same 'quarto render' again (even without changing anything), I get:
ERROR: The process cannot access the file because it is being used by another process. (os error 32): 

If I go through the same process with RStudio, I have not problem rendering multiple times. However, if also open the project (which is also a Github repository) in Positron and then try to render it more than once in RStudio, I get the same error. It seems Positron (or an extension, like Github?) is interfering with Quarto cleanup.

I have a fairly simple project, just 11 qmd files, minimal code chunks, for a basic website: https://github.com/andreashandel/SISMID2025

Some positron info:
Positron Version: 2025.06.0 (system setup) build 167
Code - OSS Version: 1.100.0
Commit: e7e5d5590adc6bf7b760f837051cd2f57ea9e3dc
Date: 2025-05-29T21:27:54.586Z
Electron: 34.5.1
Chromium: 132.0.6834.210
Node.js: 20.19.0
V8: 13.2.152.41-electron.0
OS: Windows_NT x64 10.0.26100

**Reply by cderv** [reply]
@andreashandel do you have some error printed ? Is there a stack trace ? 
Can you opt in verbose mode to give us more details on the issue ? 

Also does running Quarto render through the command palette gives you the same issue ? 

With this kind of issue, it could also depend on which extension you have in your Positron setup. If one is scanning all files and does some locking on the file, this will cause a file locking issue on Windows. 

I'll be interested with the log results as I did not encountered that recently on Windows, except with dev version. And this would reactivate interest in 
- https://github.com/quarto-dev/quarto-cli/pull/10629
Thanks

**Reply by andreashandel** [reply]
@cderv here's a bit more information:

TLDR: The Windsurf Extension that I use for AI support is the problem.

Details:

I tried to replicate the problem on a different, very similarly configured computer. It "works", i.e. I get the same error doing the following:
- Cloning the above referenced repo to my local Windows machine.
- Opening the repo with Positron (build 167), running Quart 1.7.32.
- Going into the terminal and typing 'quarto render'. Below is the full error message, labelled with 1. 

Curiously, sometimes (I don't know when/why), the same error message shows up right at the beginning of the rendering process. I placed one of these below, labelled with 2.

I do get the same error when I run Quarto through the command palette (though I got it to work once, but then I tried again and the error came back).

The extensions I have running (other than those that come pre-installed) that might be relevant: Windsurf, GitHub PR (the other computer has more, but the one also producing error messages just has those currently running). 

I tried to turn the extensions off. And it seems that the **Windsurf Plugin is the problem**. After I turn it off, I can render on both computers without getting the error message (tried multiple times on each machine). Then I turned it back on, the error message comes back.

So I guess my simple fix is to temporarily disable Windsurf AI support before rendering. And long-term waiting for the built-in Positron AI assistant :)

############## Error 1
PS C:\Data\GitHub\andreashandel\SISMID2025> quarto render
[ 1/13] Course_Communication.qmd
[ 2/13] Course_Project.qmd
[ 3/13] Course_Resources.qmd
[ 4/13] Course_Schedule.qmd
[ 5/13] Course_Syllabus.qmd
[ 6/13] andreas-presentations\building-your-brand-presentation\2025-SISMID-building-your-brand.qmd
[ 7/13] andreas-presentations\introduction-to-modeling\introduction-to-modeling-slides.qmd
[ 8/13] andreas-presentations\modeling-bacterial-infections\2025_Bacteria_Models_Slides.qmd
[ 9/13] andreas-presentations\modeling-coinfections\2025-modeling-coinfections-slides.qmd
[10/13] andreas-presentations\modeling-macroparasites\2025-modeling-macroparasites-slides.qmd
[11/13] andreas-presentations\modeling-vaccines-drugs\2025_Vaccines_Drugs_Slides.qmd
[12/13] andreas-presentations\modeling-viral-infections\2025_Virus_Models_Slides.qmd
[13/13] index.qmd

ERROR: The process cannot access the file because it is being used by another process. (os error 32): rename 'C:\Data\GitHub\andreashandel\SISMID2025\andreas-presentations\modeling-viral-infections\2025_Virus_Models_Slides.html' -> 'C:\Data\GitHub\andreashandel\SISMID2025\docs\andreas-presentations\modeling-viral-infections\2025_Virus_Models_Slides.html'

Stack trace:
    at Object.renameSync (ext:deno_fs/30_fs.js:267:3)
    at safeMoveSync (file:///C:/Program Files/Quarto/bin/quarto.js:2354:14)
    at renderProject (file:///C:/Program Files/Quarto/bin/quarto.js:87372:17)
    at eventLoopTick (ext:core/01_core.js:175:7)
    at async Command.actionHandler (file:///C:/Program Files/Quarto/bin/quarto.js:91975:32)
    at async Command.execute (file:///C:/Program Files/Quarto/bin/quarto.js:8253:13)
    at async Command.parseCommand (file:///C:/Program Files/Quarto/bin/quarto.js:8143:20)
    at async quarto (file:///C:/Program Files/Quarto/bin/quarto.js:133948:9)
    at async file:///C:/Program Files/Quarto/bin/quarto.js:133978:9
    at async file:///C:/Program Files/Quarto/bin/quarto.js:133830:20
####### End Error 1


####### Error 2
PS C:\Data\GitHub\andreashandel\SISMID2025> quarto render
ERROR: The process cannot access the file because it is being used by another process. (os error 32): remove 'C:\Data\GitHub\andreashandel\SISMID2025\docs'

Stack trace:
    at Object.removeSync (ext:deno_fs/30_fs.js:250:3)
    at safeRemoveSync (file:///C:/Program Files/Quarto/bin/quarto.js:2369:14)
    at removeIfExists (file:///C:/Program Files/Quarto/bin/quarto.js:5009:9)
    at renderProject (file:///C:/Program Files/Quarto/bin/quarto.js:87257:17)
    at eventLoopTick (ext:core/01_core.js:175:7)
    at async Command.actionHandler (file:///C:/Program Files/Quarto/bin/quarto.js:91975:32)
    at async Command.execute (file:///C:/Program Files/Quarto/bin/quarto.js:8253:13)
    at async Command.parseCommand (file:///C:/Program Files/Quarto/bin/quarto.js:8143:20)
    at async quarto (file:///C:/Program Files/Quarto/bin/quarto.js:133948:9)
    at async file:///C:/Program Files/Quarto/bin/quarto.js:133978:9
####### End Error 2

**Reply by cderv** [reply]
As this is a locked file issue, this is definitely possible that Windsurf AI is the culprit by accessing the file at the same time as Quarto. 

Positron Assistant is now in Preview if you want to give it a try: https://positron.posit.co/assistant.html

**Reply by michalovadek** [reply]
just a +1 on this: I get the same error on Windows without any AI assistant on but only in the post-render cleanup of temp files (e.g. AppData\Local\Temp\quarto-session2528fa37a846f479). This happens consistently across Quarto projects when rendered in Positron but not in RStudio

**Reply by cderv** [reply]
@michalovadek can you provide a project where this happens consistently for you ? And also system information so we can reproduce in a similar environment ? 

Which version of Positron and Quarto extension are you using ? 

I use this IDE too and don't have the issue. So what we can get from the "this is windsurf AI extension" is that some VSCODE extension may interact with filesystem at the same time as quarto cleaning. 

I'll try to build you a version with the PR I tried so that you can test.

**Reply by michalovadek** [reply]
I get `Error occurred during cleanup: Error: The process cannot access the file because it is being used by another process. (os error 32): remove 'C:\Users\me\AppData\Local\Temp\quarto-session19b9f2822e75451a'` from the following reproducible flow on my machine:

1. create new, empty folder
2. open folder in Positron
3. run `quarto::quarto_create_project(name = ".", title = "test-os-error")` and confirm (Y)
4. run `quarto::quarto_render()`

so just to be clear, the file gets rendered fine, just the temp files are not cleaned up.

quarto_version: ‘1.7.17’

Positron About:

Positron Version: 2025.10.0 (system setup) build 199
Code - OSS Version: 1.103.0
Commit: 81d0c98dbf719e29097b9c700877398d7af13138
Date: 2025-10-09T17:53:25.533Z
Electron: 37.2.3
Chromium: 138.0.7204.100
Node.js: 22.17.0
V8: 13.8.500258-electron.0
OS: Windows_NT x64 10.0.22631

Sys.info:

       sysname        release        version       nodename        machine          login           user effective_user 
     "Windows"       "10 x64"  "build 22631"   "xx"       "x86-64"      "x"      "x"      "x" 
     
sessionInfo:

R version 4.4.3 (2025-02-28 ucrt)
Platform: x86_64-w64-mingw32/x64
Running under: Windows 11 x64 (build 22631)

Matrix products: default

locale:
[1] LC_COLLATE=English_United Kingdom.utf8  LC_CTYPE=English_United Kingdom.utf8    LC_MONETARY=English_United Kingdom.utf8 LC_NUMERIC=C                            LC_TIME=English_United Kingdom.utf8    

time zone: Europe/London
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

loaded via a namespace (and not attached):
 [1] digest_0.6.37     R6_2.6.1          later_1.4.2       fastmap_1.2.0     xfun_0.52         knitr_1.50        htmltools_0.5.8.1 rmarkdown_2.29    lifecycle_1.0.4   ps_1.9.1          cli_3.6.5         processx_3.8.6    compiler_4.4.3   
[14] rstudioapi_0.17.1 tools_4.4.3       quarto_1.5.1      evaluate_1.0.3    yaml_2.3.10       Rcpp_1.1.0        rlang_1.1.6       jsonlite_2.0.0

**Comment by xx02al** [comment]
**Quarto 1.9.8 errors when previewing R package vignettes**

Hey @cderv, when we try to preview a quarto vignette we get

```
ERROR: The process cannot access the file because it is being used by another process. (os error 32): remove 'C:\Users\user\path\to\pkg\vignettes\.quarto'
```

Here's how you can reproduce it:

```
library(devtools)
create_package("~/path/to/pkg")
use_vignette("my-vignette")
```

In the `DESCRIPTION` file set

```
VignetteBuilder:
    quarto
```

Modify the vignette to at least be quarto format. For example copy the head of https://github.com/r-lib/pkgdown/blob/main/vignettes/quarto.qmd and change the file format from `.Rmd` to `.qmd`.

Trying to preview the quarto vignette errors as described.

For the purposes of this report, I tried out several Quarto projects (book, website, standard) and can confirm that everything works perfectly with the templates.

**Reply by cderv** [reply]
Thanks for the report. This is quite puzzling. Is this happening in an IDE ? or at terminal or R console ? 

> ERROR: The process cannot access the file because it is being used by another process. (os error 32)

This is happening when another process tries to do something with the same file on Windows. And sometimes it happens with VSCODE extension, or other tooling that polls. 

Still trying to think how we can protect that in Quarto.

**Reply by xx02al** [reply]
I am working in Positron (2025.10.1 build 4) and this error appears in the terminal. It appears also if I follow: `remove 'C:\Users\user\path\to\pkg\vignettes\.quarto'`, i.e., delete that, and I am also not deliberately doing anything with contents of the `\.quarto` folder.

**Reply by cderv** [reply]
And outside of Positron, you probably do not have this problem right ? 

One recent user found that is windsurf extension was the problem (https://github.com/quarto-dev/quarto-cli/discussions/10775#discussioncomment-13836245) 

So I don't have a fix right now, except thinking about how we can protect against that. 
But this is a Windows OS file locking problem that we need to handle.

**Reply by xx02al** [reply]
Yes, I could go to RStudio and render the vignette properly.
I don't have the windsurf extension installed.

**Reply by andreashandel** [reply]
@cderv I just want to add that I also regularly get this kind of error message if I operate inside a folder that is controlled by some sync service (e.g. Dropbox, OneDrive). I always pause syncing when I know I need to do a bunch of Quarto rendering. I had turned it off but still got the error, that's why I initially posted the question above, until I realized that Windsurf is apparently another program that monitors and 'locks' files. It seems there are a number of programs that 'grab on' to files produced during the whole quarto render pipeline, which can cause issues.

**Reply by cderv** [reply]
> It seems there are a number of programs that 'grab on' to files produced during the whole quarto render pipeline, which can cause issues.

Yes - it seems to be a Windows limitations with file operation, and Deno filesystem operation does not seems to handle locking correctly on their own. I don't know if we do something wrong, or if we are just not helped by Deno and must find a way to protect against that. #10629 is maybe helpful - but if did not felt that efficient. I'll try to build a new special version for you to try maybe
